pkg/collector/distinct_value_collector.go

```diff
@@ -73,26 +90,32 @@ func (d *DistinctValue[T]) Values() []T {
 	return ss
 }
 
-// Exceeded indicates if some values were lost because the maximum size limit was met.
+// Exceeded indicates that we have exceeded the limit
+// can be used to stop early and to avoid collecting further values
 func (d *DistinctValue[T]) Exceeded() bool {
-	d.mtx.RLock()
-	defer d.mtx.RUnlock()
-	return d.totalLen > d.currLen
+	d.mtx.Lock()
+	defer d.mtx.Unlock()
+	return d.limExceeded
 }
 
-// TotalDataSize is the total size of all distinct strings encountered.
-func (d *DistinctValue[T]) TotalDataSize() int {
-	d.mtx.RLock()
-	defer d.mtx.RUnlock()
-	return d.totalLen
+// Size is the total size of all distinct items collected
+func (d *DistinctValue[T]) Size() int {
+	d.mtx.Lock()
+	defer d.mtx.Unlock()
+	return d.currLen
 }
 
 // Diff returns all new strings collected since the last time diff was called
+// returns nil if diff is not enabled
 func (d *DistinctValue[T]) Diff() []T {
+	if !d.diffEnabled {
+		return nil
+	}
+
 	ss := make([]T, 0, len(d.new))
```

Comment:
this was not under lock before, but i think that's a minor bug. we should move it down after the lock is taken